addons/pos_self_order/static/src/router.js

```diff
@@ -0,0 +1,94 @@
+/** @odoo-module */
+
+import { Component, useState, useSubEnv, xml } from "@odoo/owl";
+import { escapeRegExp } from "@web/core/utils/strings";
+import { browser } from "@web/core/browser/browser";
+import { zip } from "@web/core/utils/arrays";
+
+function parseParams(matches, paramSpecs) {
+    return Object.fromEntries(
+        zip(matches, paramSpecs).map(([match, paramSpec]) => {
+            const { type, name } = paramSpec;
+            switch (type) {
+                case "int":
+                    return [name, parseInt(match)];
+                case "string":
+                    return [name, match];
+                default:
+                    throw new Error(`Unknown type ${type}`);
+            }
+        })
+    );
+}
+export class Router extends Component {
+    static template = xml`<t t-slot="{{state.activeSlot}}" t-props="state.slotProps"/>`;
+    setup() {
+        useSubEnv({ navigate: this.navigate.bind(this) });
+        this.state = useState({
+            activeSlot: "default",
+            slotProps: {},
+        });
+        window.addEventListener("self_order_navigate", () => {
+            this.matchURL();
+        });
+        // this is needed for the back button to work
+        window.addEventListener("popstate", (event) => {
+            event.preventDefault();
+            this.matchURL();
+        });
+        this.routes = Object.keys(this.props.slots).map((route) => {
+            const paramStrings = route.match(/\{\w+:\w+\}/g);
+            if (!paramStrings) {
+                return { route, paramSpecs: [], regex: new RegExp(`^${route}$`) };
+            }
+            const paramSpecs = paramStrings.map((paramString) => {
+                const [, type, name] = paramString.match(/(\w+):(\w+)/);
+                return { type, name };
+            });
+            const regex = new RegExp(
+                `^${route
+                    .split(/\{\w+:\w+\}/)
+                    .map((part) => escapeRegExp(part))
+                    .join("(.*)?")}$`
```

Comment:
should probably be `([^/]+)`, we don't want empty segments and they shouldn't ever contain a slash.